bin/ch-run.c

```diff
@@ -215,8 +248,15 @@ void fix_environment(struct args *args)
    for (int i = 0; args->env_deltas[i].action != END; i++) {
       char *arg = args->env_deltas[i].arg;
       if (args->env_deltas[i].action == SET_FILE) {
-         FILE *fp;
-         Tf (fp = fopen(arg, "r"), "--set-env: can't open: %s", arg);
+         FILE *fp = NULL;
+         if (strchr(arg, '=') == NULL) {
+            Tf (fp = fopen(arg, "r"), "--set-env: can't open: %s", arg);
+         }
+         else {  
+            split(&name, &new_value, arg, '=');
+            env_parse(name, new_value, args->c.expand);
```

Comment:
I think you should structure this a bit differently. Read the file and create a list of lines; *or* create a single-item list of lines. Then both lists can go through the same loop.